rollout.py

```diff
@@ -0,0 +1,377 @@
+                # It's stale; terminate it
+                rel['metadata']['rollout'] = completed_rollout
+
+    # Drop completed rollouts except for the last one
+    predicate = lambda rel: rel['metadata'].get('rollout') == completed_rollout
+    for rel in islice(filter(predicate, reversed(info['releases'])), 1, None):
+        del rel['metadata']['rollout']
+
+    # Drop releases that have no rollout, barrier, or deadend
+    def predicate(rel):
+        for section in 'barrier', 'deadend', 'rollout':
+            if section in rel['metadata']:
+                return True
+        return False
+    info['releases'] = [rel for rel in info['releases'] if predicate(rel)]
+
+
+def report(info):
+    '''Summarize the latest rollout.'''
+    stream = info["stream"]
+    if not info["releases"]:
+        print(f"{stream} has no rollouts")
+        return
+    release = info["releases"][-1]
+    version = release["version"]
+    rollout = release["metadata"].get("rollout", None)
+    if not rollout:
+        print(f"latest entry {version} on {stream} is not a rollout")
+        return
+
+    start_percentage = rollout["start_percentage"]
+    # totally just going to ignore floating-point concerns here
+    if int(start_percentage * 100) == 100:
+        print(f"{stream} rollout of {version} at 100%")
+        return
+    ts = datetime.fromtimestamp(rollout["start_epoch"], timezone.utc)
+    raleigh_ts = ts.astimezone(dateutil.tz.gettz("America/Toronto"))
+    berlin_ts = ts.astimezone(dateutil.tz.gettz("Europe/Berlin"))
+    mins = rollout["duration_minutes"]
+    hrs = mins / 60.0
+    ts_now = datetime.now(timezone.utc)
+    if ts_now > ts:
+        delta_str = str(ts_now - ts).split(".")[0]
+        delta_str = f"{delta_str} ago"
+    else:
+        delta_str = str(ts - ts_now).split(".")[0]
+        delta_str = f"in {delta_str}"
+    print(f"{stream}")
+    print(f"    version: {version}")
+    print(f"    start: {ts} UTC ({delta_str})")
+    print(f"           {raleigh_ts} Raleigh/New York/Toronto")
+    print(f"           {berlin_ts} Berlin/France/Poland")
+    print(f"    duration: {mins}m ({hrs}h)")
+
+
+def path(stream):
+    '''Get the relative path of an update JSON file for a stream.'''
+    return f'updates/{stream}.json'
+
+
+def _do_add(args):
+    selftest()
+    info = load(path(args.stream))
+    clean(info)
+    add(info, args.version, args.start, args.duration, barrier=args.barrier,
+            deadend=args.deadend)
+    report(info)
+    save(path(args.stream), info)
+
+
+def _do_clean(args):
+    selftest()
+    for stream in args.stream:
+        info = load(path(stream))
+        clean(info)
+        save(path(stream), info)
+
+
+def _do_print(args):
+    for stream in args.stream:
+        info = load(path(stream))
+        report(info)
+
+
+def _main():
+    parser = argparse.ArgumentParser(description='Manage rollouts.')
+    subcommands = parser.add_subparsers(title='subcommands', required=True)
```

Comment:
This needs a `metavar`, otherwise argparse crashes when it tries to print the help string when calling the root of the script:

```
$ ./rollout.py
Traceback (most recent call last):
  File "/home/jlebon/Code/github.com/coreos/fedora-coreos-streams/./rollout.py", line 377, in <module>
    _main()
  File "/home/jlebon/Code/github.com/coreos/fedora-coreos-streams/./rollout.py", line 187, in _main
    args = parser.parse_args()
  File "/usr/lib64/python3.9/argparse.py", line 1818, in parse_args
    args, argv = self.parse_known_args(args, namespace)
  File "/usr/lib64/python3.9/argparse.py", line 1851, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/usr/lib64/python3.9/argparse.py", line 2089, in _parse_known_args
    ', '.join(required_actions))
TypeError: sequence item 0: expected str instance, NoneType found
```

Reply:
`metavar` overrides the automatic listing of subcommands in the `--help` text.  Setting `dest` instead, as recommended [here](https://bugs.python.org/issue29298).